src/dos/dos_code_page.cpp

```diff
@@ -0,0 +1,2223 @@
+		return false;
+	}
+
+	// If we are that close to the end of the content, it is very unlikely
+	// the data represent useful 'CodePageEntryHeader'. To be sure, check
+	// the next two bytes, which should represent the structure size.
+	if (remaining < 2) {
+		return true;
+	}
+	const size_t value = content[offset] + (content[offset + 1] << 8);
+	if (value > remaining) {
+		return true;
+	}
+
+	return false;
+}
+
+std::vector<uint8_t> CpiReader::ReadBlob(const uint32_t length)
+{
+	const auto idx = offset;
+
+	offset += length;
+	if (IsOffsetBeyondContent()) {
+		WarnDataOutOfBounds(idx, offset);
+		return {};
+	}
+
+	return std::vector<uint8_t>(content.begin() + idx,
+	                            content.begin() + idx + length);
+}
+
+std::optional<FontFileHeader> CpiReader::ReadFontFileHeader()
+{
+	FontFileHeader result = {};
+
+	result.struct_offset = offset;
+
+	result.id_byte   = InternalReadByte();
+	result.id_string = InternalReadString(7);
+	SkipBytes(8);
+	result.num_pointers = InternalReadShort();
+	result.pointer_type = InternalReadByte();
+	result.fih_offset   = InternalReadLong();
+
+	if (IsOffsetBeyondContent()) {
+		WarnDataOutOfBounds(result.struct_offset, offset, result.StructName);
+		return {};
+	}
+	return result;
+}
+
+std::optional<FontFileExtendedHeader> CpiReader::ReadFontFileExtendedHeader()
+{
+	FontFileExtendedHeader result = {};
+
+	result.struct_offset = offset;
+
+	result.fonts_per_codepage = InternalReadByte();
+
+	for (size_t idx = 0; idx < result.fonts_per_codepage; ++idx) {
+		result.cell_size.push_back(InternalReadByte());
+	}
+
+	for (size_t idx = 0; idx < result.fonts_per_codepage; ++idx) {
+		result.dfd_offset.push_back(InternalReadLong());
+	}
+
+	if (IsOffsetBeyondContent()) {
+		WarnDataOutOfBounds(result.struct_offset, offset, result.StructName);
+		return {};
+	}
+	return result;
+}
+
+std::optional<FontInfoHeader> CpiReader::ReadFontInfoHeader()
+{
+	FontInfoHeader result = {};
+
+	result.struct_offset = offset;
+
+	result.num_code_pages = InternalReadShort();
+
+	if (IsOffsetBeyondContent()) {
+		WarnDataOutOfBounds(result.struct_offset, offset, result.StructName);
+		return {};
+	}
+	return result;
+}
+
+std::optional<CodePageEntryHeader> CpiReader::ReadCodePageEntryHeader()
+{
+	CodePageEntryHeader result = {};
+
+	result.struct_offset = offset;
+
+	result.cpeh_size        = InternalReadShort();
+	result.next_cpeh_offset = InternalReadLong();
+	result.device_type      = InternalReadShort();
+	result.device_name      = InternalReadString(8);
+	result.code_page        = InternalReadShort();
+	SkipBytes(6);
+
+	// Strip leading device name spaces
+	while (result.device_name.ends_with(" ")) {
+		result.device_name.pop_back();
+	}
+
+	// According to https://www.seasip.info/DOS/CPI/cpi.html:
+	// - the structure is normally 28 bytes
+	// - many tools simply ignore the size
+	// - sometimes (in old files) the structure is 26 bytes due to offset
+	//   being stored on 2 bytes
```

Comment:
The comment is good, but extract the 26 and 28 magic numbers to constants with sensible names.

Reply:
Done.